moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -117,18 +156,78 @@ abstract T newInstance() throws
       final Method newInstance = ObjectInputStream.class.getDeclaredMethod(
           "newInstance", Class.class, Class.class);
       newInstance.setAccessible(true);
-      return new ClassFactory<T>() {
+      ClassFactory<T> factory = new ClassFactory<T>() {
         @SuppressWarnings("unchecked")
-        @Override public T newInstance() throws InvocationTargetException, IllegalAccessException {
+        @Override
+        public T newInstance() throws InvocationTargetException, IllegalAccessException {
           return (T) newInstance.invoke(null, rawType, Object.class);
         }
-        @Override public String toString() {
+
+        @Override
+        public String toString() {
           return rawType.getName();
         }
       };
+      unsafeFactories.put(rawType, factory);
+      return factory;
     } catch (Exception ignored) {
     }
 
     throw new IllegalArgumentException("cannot construct instances of " + rawType.getName());
   }
+
+  private static <T> ClassFactory<T> createAnnotationClassFactory(final Class<?> rawType, final Constructor<?> constructor) {
+    return new ClassFactory<T>() {
+            @SuppressWarnings("unchecked")
+            @Override
+            T newInstance() throws InvocationTargetException, IllegalAccessException, InstantiationException {
+              Class<?>[] parameterTypes = constructor.getParameterTypes();
+              Object[] args = new Object[parameterTypes.length];
+              for (int j = 0; j < parameterTypes.length; j++) {
+                Class<?> parameterType = parameterTypes[j];
+                // Default values according to
+                // https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
+                if (parameterType == byte.class || parameterType == Byte.class) {
```

Comment:
Good point. Perhaps my [suggestion](https://github.com/square/moshi/pull/266#issuecomment-284547299) to make the `ClassFactory` pluggable may be a better and more flexible option?